### PR TITLE
WIP Separate sector data from repo

### PR DESCRIPTION
**Pull request by ZenGround0**
- separate repo interface into Repo and SectorRepo
- fsrepo handles repo data at the new <root>/repo path

### CODE REVIEW

**Comment by ZenGround0** [comment]
@anorth @shannonwells it seemed like all of this was covered heavily enough by tests but please let me know if there are areas that you would like to see tested.

**Comment by ZenGround0** [comment]
@laser just added you to make sure I'm not reking any rust-fil-proofs guarantees, no need to look at the whole thing.

File: repo/fsrepo_test.go [hunk 1]
```diff
@@ -178,9 +186,10 @@ func TestFSRepoRoundtrip(t *testing.T) {
 	cfg := config.NewDefaultConfig()
 	cfg.API.Address = "foo" // testing that what we get back isnt just the default
 	assert.NoError(err, InitFSRepo(dir, cfg))
-
+	fmt.Printf("expected dir %v\n", dir)
```
Comment (shannonwells) [inline repo/fsrepo_test.go]: remove debug print

File: repo/fsrepo_test.go [hunk 2]
```diff
@@ -178,9 +186,10 @@ func TestFSRepoRoundtrip(t *testing.T) {
...
 	r, err := OpenFSRepo(dir)
 	assert.NoError(err)
+	fmt.Printf("r.repoPath: %s, r.rootPath: %s\n", r.repoPath, r.rootPath)
```
Comment (shannonwells) [inline repo/fsrepo_test.go]: same

File: repo/fsrepo.go [hunk 3]
```diff
@@ -493,12 +518,17 @@ func (r *FSRepo) SetAPIAddr(maddr string) error {
 	return nil
 }
 
+func APIAddrOfRoot(rootDir string) (string, error) {
```
Comment (shannonwells) [inline repo/fsrepo.go]: There were some other checks in this code you abstracted -- were they not necessary?
```
rawPath := filepath.Join(filepath.Clean(repoDir), repo.APIFile)	
		apiFilePath, err := homedir.Expand(rawPath)	
		if err != nil {	
			return "", errors.Wrap(err, fmt.Sprintf("can't resolve local repo path %s", rawPath))	
		}
```

**Comment by laser** [comment]
@ZenGround0 

> @laser just added you to make sure I'm not reking any rust-fil-proofs guarantees, no need to look at the whole thing.

I don't see anything that's going to wreck us. Thanks for asking.

**Comment by ZenGround0** [comment]
@travisperson tagging to make sure I didn't ruin any iptb stuff.  No need to look at the whole thing.

File: repo/fsrepo.go (hunk 3 above)
Comment (ZenGround0) [reply repo/fsrepo.go]: I think this is a bug, thanks for catching.

File: repo/fsrepo.go [hunk 4]
```diff
@@ -67,41 +69,42 @@ type FSRepo struct {
...
+	rootDir = GetHomeDir(rootDir)
+	if err := InitFSRepo(rootDir, cfg); err != nil {
 		return nil, err
 	}
-	return OpenFSRepo(repoDir)
+	return OpenFSRepo(rootDir)
 }
 
 // OpenFSRepo opens an already initialized fsrepo at the given path
 func OpenFSRepo(p string) (*FSRepo, error) {
```
Comment (anorth) [inline repo/fsrepo.go]: Change `p` and the comment to clearly refer to `homePath` not the repo path.

Possibly better would be to take a repo path here and push the expansion and joining up to the caller.

File: repo/fsrepo.go [hunk 5]
```diff
@@ -169,12 +172,18 @@ func (r *FSRepo) loadFromDisk() error {
 
 // InitFSRepo initializes an fsrepo at the given path using the given configuration
 func InitFSRepo(p string, cfg *config.Config) error {
-	expath, err := homedir.Expand(p)
+	homePath, err := homedir.Expand(p)
```
Comment (anorth) [inline repo/fsrepo.go]: Ditto

File: repo/fsrepo.go [hunk 6]
```diff
@@ -458,19 +467,35 @@ func fileExists(file string) bool {
...
+func (r *FSRepo) SealedDir() (string, error) {
+	// Default is to keep sector data alongside node repo data.
+	if r.cfg.SectorBase.RootDir == "" {
+		return path.Join(r.homePath, "sectors/sealed"), nil
+	}
+	// If not using default check for writeability
+	if err := checkWritable(r.cfg.SectorBase.RootDir); err != nil {
+		return "", nil
+	}
+	return path.Join(r.cfg.SectorBase.RootDir, "sealed"), nil
```
Comment (anorth) [inline repo/fsrepo.go]: As noted elsewhere, remove this responsibility from "fsrepo". I would suggest the permission checks are out of place here, too.

**Comment by ZenGround0** [comment]
@anorth I've pushed us to use a model where users specify the repo dir directly as opposed to the home dir.  If this is still has not gone far enough lets sync on what we want for users when it comes to default directories and configurations of home/repo/sector dirs.

File: paths/paths.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package paths
```
Comment (anorth) [inline paths/paths.go]: This probably shouldn't be a top-level package, but thanks for breaking it out. I don't have a concrete better suggestion at the moment.

File: repo/mem.go [hunk 8]
```diff
@@ -114,26 +90,9 @@ func (mr *MemRepo) Version() uint {
...
-func (mr *MemRepo) SealedDir() string {
-	return mr.sealedDir
-}
-
-// CleanupSectorDirs removes all sector directories and their contents.
-func (mr *MemRepo) CleanupSectorDirs() {
-	os.RemoveAll(mr.StagingDir()) // nolint: errcheck
-	os.RemoveAll(mr.SealedDir())  // nolint:errcheck
-}
-
```
Comment (anorth) [inline repo/mem.go]: 👍

**Pull request merged**